astropy/modeling/tests/test_quantities_fitting.py

```diff
@@ -0,0 +1,143 @@
+    """
+    Raise an error if the model has units but the data doesn't
+    """
+    g_init = Gaussian1D(amplitude=1. * u.mJy, mean=3 * u.cm, stddev=2 * u.mm)
+    fit_g = fitting.LevMarLSQFitter()
+
+    with pytest.raises(UnitsError) as exc:
+        fit_g(g_init, [1, 2, 3], [4, 5, 6])
+    assert exc.value.args[0] == ("Units of input 'x', (dimensionless), does not "
+                                 "match required units for model input, cm (length)")
+
+    with pytest.raises(UnitsError) as exc:
+        fit_g(g_init, [1, 2, 3] * u.m, [4, 5, 6])
+    assert exc.value.args[0] == ("Units of input 'y', (dimensionless), does not "
+                                 "match required units for model output, Jy")
+
+
+@pytest.mark.xfail
+def test_fitting_missing_model_units():
+    """
+    Raise an error if the data has units but the model doesn't.
+    """
+
+    # TODO: determine whether this breaks backward-compatibility.
+
+    g_init = Gaussian1D(amplitude=1., mean=3, stddev=2)
+    fit_g = fitting.LevMarLSQFitter()
+
+    with pytest.raises(UnitsError) as exc:
+        fit_g(g_init, [1, 2, 3] * u.m, [4, 5, 6] * u.Jy)
+    assert exc.value.args[0] == ("Units of input 'x', m (length), does not "
+                                 "match required units for model input, "
+                                 "(dimensionless)")
+
+    g_init = Gaussian1D(amplitude=1., mean=3 * u.m, stddev=2 * u.m)
+    fit_g = fitting.LevMarLSQFitter()
+
+    with pytest.raises(UnitsError) as exc:
+        fit_g(g_init, [1, 2, 3] * u.m, [4, 5, 6] * u.Jy)
+    assert exc.value.args[0] == ("Units of input 'y', Jy, does not "
+                                 "match required units for model output, "
+                                 "(dimensionless)")
+
+
+@pytest.mark.xfail
+def test_fitting_incompatible_units():
+    """
+    Raise an error if the data and model have incompatible units
+    """
+
+    g_init = Gaussian1D(amplitude=1. * u.Jy, mean=3 * u.m, stddev=2 * u.cm)
+    fit_g = fitting.LevMarLSQFitter()
+
+    with pytest.raises(UnitsError) as exc:
+        fit_g(g_init, [1, 2, 3] * u.Hz, [4, 5, 6] * u.Jy)
+    assert exc.value.args[0] == ("Units of input 'x', Hz (frequency), does not "
+                                 "match required units for model input, "
+                                 "m (length)")
+
+
+@pytest.mark.xfail
+def test_fitting_with_equivalencies():
+    """
+    Check that equivalencies are correctly taken into account in the fitting.
+    """
+
+    # A simple test with the spectral equivalency
+
+    g_init = Gaussian1D(amplitude=1. * u.Jy, mean=3 * u.m, stddev=2 * u.cm)
+    g_init.input_equivalencies = u.spectral()
+
+    fit_g = fitting.LevMarLSQFitter()
+
+    g = fit_g(g_init, [1, 2, 3] * u.Hz, [4, 5, 6] * u.Jy)
+
+    # TODO: numerical test of results
+
+    # A more complex test with an equivalency that depends on x values
+    g_init.output_equivalencies = u.brightness_temperature, (3e-5 * u.sr, 'x')
```

Comment:
Indeed this is needed because 'x' is not known until the model is evaluated. I described this a little more in ``test_output_units_equivalencies_with_parameters``. There may indeed be better ways to specify this syntax-wise.